Increase limits of guest purchase statistics and prevent them from overflowing

matheusvb3 opened the pull request:
Fixes #21859 and #21333

While I was there I decided to tidy up some stuff in `Guest::DecideAndBuyItem`.

I would greatly appreciate if someone could help test this to make sure there are no side effects.

Gymnasiast commented on src/openrct2/entity/Guest.cpp:
```diff
@@ -1656,32 +1656,38 @@ bool Guest::DecideAndBuyItem(Ride& ride, const ShopItem shopItem, money64 price)
+            Photo4RideRef = ride.id;
+            break;
+        default:
+            break; // Fallback
```
Comments should tell us something. This just states the obvious.

```suggestion
            break;
```

Gymnasiast commented on src/openrct2/entity/Guest.cpp:
```diff
@@ -1696,27 +1702,27 @@ bool Guest::DecideAndBuyItem(Ride& ride, const ShopItem shopItem, money64 price)
     }
 
-    if (shopItemDescriptor.IsFood())
+    if (shopItemDescriptor.IsFood() && AmountOfFood < std::numeric_limits<decltype(AmountOfFood)>::max())
```
We already have functions for this called AddClamp. Please use these.

matheusvb3 commented:
@Gymnasiast I've addressed your comments, tried to make the AddClamp usage modular so it'll adapt to whatever data type the guest stats are (thus if it ever gets changed we'll only need to change it in `Guest.h`), is that OK?

matheusvb3 commented:
I've updated strings 2297-2304 (X amount spent on Y thing) to use COMMA32 instead of COMMA16, otherwise there would an overflow in the statistics tab I hadn't actually caught. Is there any way we apply this change to other languages?

Gymnasiast commented:
The other languages can be updated with a pull request to Localisation.

duncanspumpkin reviewed (commented):
Will need network version bump and minimum park file version bump.

matheusvb3 commented:
Oh I did something in Github Desktop and I'm not quite sure if I like it. Hopefully shouldn't cause any problems.....

> Will need network version bump and minimum park file version bump.

Could you instruct me on how to do that, please?

AaronVanGeffen commented:
@duncanspumpkin Not sure the park file version needs incrementing. Integers are already expanded by the OrcaStream class, anyway, aren't they?

duncanspumpkin commented:
> @duncanspumpkin Not sure the park file version needs incrementing. Integers are already expanded by the OrcaStream class, anyway, aren't they?

They are expanded but it performs a size check which will cause the file to fail to load. That is why all size increases must be a new version.

duncanspumpkin commented:
@matheusvb3 you'll need to increment the number here https://github.com/OpenRCT2/OpenRCT2/blob/5bec62c540b8bea53452cf2ccf33c29711f8d49b/src/openrct2/park/ParkFile.h#L17 and the current version just a line above. The network version can be found in Network.cpp just increase the number by 1.

tupaschoal commented:
@matheusvb3 I've done the park file and network bumps for you, check the second commit.

I also didn't understand what you had done with the merge, but I fixed the commit history again.

matheusvb3 commented:
Thank you!!! I guess I tried to sync my fork with the develop branch but I didn't expect all that to happen 😅

Since changing the guest stats requires incrementing the network and park file versions I think it might be worth waiting a little longer because I've found other issues with staff stats. I've pinged @Gymnasiast on Discord to talk about it... Once a fix is found we can merge both PRs at once to avoid incrementing twice, if that's desired.

Gymnasiast commented:
@matheusvb3 It depends on what you found in terms of staff stats. If it’s just a matter of adding clamps, it won’t require a new park file version.

Whatever the extent of the staff stats, I think it’s OK to address them in this PR.

matheusvb3 commented:
> If it’s just a matter of adding clamps, it won’t require a new park file version.

It's a little more complicated... I think I'll make an issue to discuss it as soon as I have the time.

matheusvb3 commented:
So I spoke with @duncanspumpkin in https://github.com/OpenRCT2/OpenRCT2/issues/23698 and we found a solution but I think it's better if I split it into two other PRs, one to clamp the staff stats increment and another to prevent the strings in the staff stats tab from overflowing. I would add the clamp in this PR, but for some reason my branch is completely borked in Github Desktop and if I push it'll add a bunch of commits I didn't actually make (happened before and @tupaschoal had to help me with it), sorry but I really don't know what I did wrong here...

duncanspumpkin commented:
> So I spoke with @duncanspumpkin in #23698 and we found a solution but I think it's better if I split it into two other PRs, one to clamp the staff stats increment and another to prevent the strings in the staff stats tab from overflowing. I would add the clamp in this PR, but for some reason my branch is completely borked in Github Desktop and if I push it'll add a bunch of commits I didn't actually make (happened before and @tupaschoal had to help me with it), sorry but I really don't know what I did wrong here...

GitHub desktop doesn't like rewriting history. You would just need to accept the history from upstream normally that would be done by force reseting but I don't know what GitHub desktop would call that. From command line it would be something like `git reset --hard <repo>/<branch>` assuming you had already done a `git fetch <repo>`

matheusvb3 commented:
Oh I did it again..... Tried running what you suggested but it seems to haven't worked 🤡

Since the guest stats are uint16 instead of 32 (like those of staff) I can't do AddClamp the same way I did in https://github.com/OpenRCT2/OpenRCT2/pull/23719 unfortunately...

tupaschoal commented:
I fixed it again. Don't use GH Desktop for merging, what you need is to use the CLI, with basically two commands:
```
git pull --rebase upstream develop
git push -f origin
```

The first will rebase your changes on top of the current develop. Most of the times it will finish without any problem, but if there is a merge conflict it will stop and ask for you to solve before continuing (`git rebase --continue`).
The second will just force the history back onto your branch, rewriting it.

It's documented in here: https://github.com/OpenRCT2/OpenRCT2/wiki/Rebase-and-Sync-fork-with-OpenRCT2

astraylife1 commented:
Keep open
